### ISSUE TITLE
Crash on import in embedded Python

### ISSUE THREAD

**Issue by mmmm1998** [matplotlib/matplotlib#14586]
### Bug report

**Bug summary**

Importing `pyplot` from matplotlib in my C++ with embedded python cause UnicodeDecodeError (traceback below).
Another modules haven't this problem, even `matplotlib`.
I see, that problem is that I used embedded python command
````C++
Py_CompileStringObject(const char *str, PyObject *filename, int start, PyCompilerFlags *flags, int optimize)
````
with `filename` points to existing source file, but this is not `.py` file, this actually zip archive with code (my program packes code to archive).
According [documentation](https://docs.python.org/3/c-api/veryhigh.html#c.Py_CompileStringObject), the `filename` used in tracebacks and SyntaxError, but it looks like pyplot tried to read the source file and failed, because it's not unicode text file but zip archive.
Without the `filename` parameter, or if it points to unexisted file, import works fine.

**Actual outcome**

````
Traceback (most recent call last):
File "/usr/lib/python3.6/tokenize.py", line 390, in find_cookie
line_string = line.decode('utf-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd9 in position 10: invalid continuation byte

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/mmmm1998/tmp.cws", line 1, in 
File "/usr/local/lib/python3.6/dist-packages/matplotlib/pyplot.py", line 71, in 
from matplotlib.backends import pylab_setup
File "/usr/local/lib/python3.6/dist-packages/matplotlib/backends/__init__.py", line 16, in 
line for line in traceback.format_stack()
File "/usr/lib/python3.6/traceback.py", line 197, in format_stack
return format_list(extract_stack(f, limit=limit))
File "/usr/lib/python3.6/traceback.py", line 211, in extract_stack
stack = StackSummary.extract(walk_stack(f), limit=limit)
File "/usr/lib/python3.6/traceback.py", line 364, in extract
f.line
File "/usr/lib/python3.6/traceback.py", line 286, in line
self._line = linecache.getline(self.filename, self.lineno).strip()
File "/usr/lib/python3.6/linecache.py", line 16, in getline
lines = getlines(filename, module_globals)
File "/usr/lib/python3.6/linecache.py", line 47, in getlines
return updatecache(filename, module_globals)
File "/usr/lib/python3.6/linecache.py", line 136, in updatecache
with tokenize.open(fullname) as fp:
File "/usr/lib/python3.6/tokenize.py", line 454, in open
encoding, lines = detect_encoding(buffer.readline)
File "/usr/lib/python3.6/tokenize.py", line 431, in detect_encoding
encoding = find_cookie(first)
File "/usr/lib/python3.6/tokenize.py", line 395, in find_cookie
raise SyntaxError(msg)
SyntaxError: invalid or missing encoding declaration for '/home/mmmm1998/tmp.cws'
````

**Expected outcome**

Succes import pyplot without errors

**Matplotlib version**

  * Operating system: Ubuntu 18.04.2 LTS
  * Matplotlib version: 2.2.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: Python 3.6.7

Python from system installation, matplotlib from pip, installed into /usr/local/

**Comment by WeatherGod** [comment]
This looks familiar, but from a different context. I think this problem was
addressed in v3.0. Have you tried a more recent version of matplotlib?

**Comment by mmmm1998** [comment]
Thanks for answer, I have updated the package and on v3.1.0 all works fine.
